src/smtsolvers/TheoryIF.cc

```diff
@@ -75,75 +78,129 @@ void CoreSMTSolver::crashTest(int rounds, Var var_true, Var var_false)
+    std::vector<int> lastAssignedLevels;
+    lastAssignedLevels.resize(splitClauses.size(), 0);
+    std::transform(splitClauses.begin(), splitClauses.end(), lastAssignedLevels.begin(), [&](auto const & clause) {
+        auto maxLevel = 0;
+        for (Lit l : clause) {
+            auto level = getVarLevel(var(l));
+            if (level > maxLevel) {
+                maxLevel = level;
+            }
+        }
+        return maxLevel;
+    });
+
+    std::vector<int> indices(splitClauses.size());
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&lastAssignedLevels](int first, int second) {
+        return lastAssignedLevels[first] < lastAssignedLevels[second];
+    });
+    return indices;
+}
+}
 
-        //printf(" -> Adding the new split\n");
+TPropRes CoreSMTSolver::handleNewSplitClauses(SplitClauses & splitClauses) {
+    vec<LitLev> deds;
+    deduceTheory(deds); // To remove possible theory deductions
+    TPropRes res = TPropRes::Undef;
+    struct PropagationData {
+        Lit lit;
+        CRef reason;
+    };
+    std::vector<PropagationData> propData;
+
+    auto processNewClause = [&] (auto const & clause) {
+        CRef cr = ca.alloc(clause, false);
+        attachClause(cr);
+        clauses.push(cr);
+        if (logsProofForInterpolation()) {
+            // MB: the proof needs to know about the new clause
+            proof->newSplitClause(cr);
+        }
+        return cr;
+    };
 
-        vec<LitLev> deds;
-        deduceTheory(deds); // To remove possible theory deductions
+    auto sortedIndices = sortByLastAssignedLevel(splitClauses, [this](Var v) { return vardata[v].level; });
 
-        Lit l1 = new_splits[0];
-        Lit l2 = new_splits[1];
+    for (int index : sortedIndices) {
+        auto & splitClause = splitClauses[index];
+        unsigned satisfied = 0;
+        unsigned falsified = 0;
+        unsigned unknown = 0;
         // MB: ensure the SAT solver knows about the variables and that they are active
-        addVar_(var(l1));
-        addVar_(var(l2));
-        assert(value(l1) == l_Undef || value(l2) == l_Undef);
-        assert(value(l1) != l_True);
-        assert(value(l2) != l_True);
-        if (value(l1) == l_Undef && value(l2) == l_Undef) {
-            // MB: allocate, attach and remember the clause - treated as original
-            // MB: TODO: why not theory clause?
-            CRef cr = ca.alloc(new_splits, false);
-            attachClause(cr);
-            clauses.push(cr);
-            if (this->logsProofForInterpolation()) {
-                // MB: the proof needs to know about the new clause
-                proof->newSplitClause(cr);
-            }
-            forced_split = ~l1;
-            return TPropRes::Decide;
+        int impliedIndex = -1;
+        for (int i = 0; i < splitClause.size(); ++i) {
+            Lit l = splitClause[i];
+            addVar_(var(l));
+            if (value(l) == l_True) { ++satisfied; }
```

Comment:
`break`?

Reply:
I thought it would be OK, but then I spotted `addVar_`. I think in our uses cases, it would not make a difference, but in general, it's better to iterate over all the literals and ensure that the variables are known to the SAT solver.

Reply:
Yeah, I missed that one! Let's keep it as it is.